### PR TITLE
Merge/staking with subassets

### PR DESCRIPTION
**Pull request by bnsports**

### CODE REVIEW

File: fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import UIKit
+
+enum StakingUnbondConfirmFlowError: Error {}
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift:3]: empty

File: fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    func didSubmitUnbonding(result: Result<String, Error>)
+
+    func provideFeeViewModel()
+    func provideAssetViewModel()
+    func provideConfirmationViewModel()
+    func refreshFeeIfNeeded()
+}
+
+protocol StakingUnbondConfirmViewModelState: StakingUnbondConfirmUserInputHandler {
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift:26]: mixed up

File: fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    ) -> StakingUnbondConfirmViewModel?
+}
+
+protocol StakingUnbondConfirmStrategy {
+    func setup()
+    func estimateFee(builderClosure: ExtrinsicBuilderClosure?, reuseIdentifier: String?)
+    func submit(builderClosure: ExtrinsicBuilderClosure?)
+}
+
+protocol StakingUnbondConfirmUserInputHandler {}
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondConfirm/Flow/StakingUnbondConfirmFlow.swift:59]: empty

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/Parachain/StakingUnbondSetupParachainStrategy.swift [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            do {
+                let response = try topDelegationsOperation.targetOperation.extractNoCancellableResultData()
+
+                guard let delegations = response else {
+                    return
+                }
+
+                self?.output?.didReceiveTopDelegations(delegations: delegations)
+            } catch {
+                print("error: ", error)
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/Parachain/StakingUnbondSetupParachainStrategy.swift:143]: debug print

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/Parachain/StakingUnbondSetupParachainViewModelState.swift [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    private var topDelegations: ParachainStakingDelegations?
+
+    var stateListener: StakingUnbondSetupModelStateListener?
+
+    let delegation: ParachainStakingDelegation
+    let candidate: ParachainStakingCandidateInfo
+    let chainAsset: ChainAsset
+    let wallet: MetaAccountModel
+    let dataValidatingFactory: StakingDataValidatingFactory
+    let callFactory: SubstrateCallFactoryProtocol = SubstrateCallFactory()
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/Parachain/StakingUnbondSetupParachainViewModelState.swift:21]: privates

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/Relaychain/StakingUnbondSetupRelaychainViewModelState.swift [hunk 6]
```diff
@@ -0,0 +1,196 @@
+import Foundation
+import BigInt
+
+final class StakingUnbondSetupRelaychainViewModelState: StakingUnbondSetupViewModelState {
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/Relaychain/StakingUnbondSetupRelaychainViewModelState.swift:4]: mixed up properties and methods

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import UIKit
+import SoraFoundation
+
+enum StakingUnbondSetupFlowError: Error {}
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift:4]: empty

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    func provideAssetViewModel()
+    func provideBondingDuration()
+
+    func provideAccountViewModel()
+    func provideCollatorViewModel()
+
+    func updateFeeIfNeeded()
+}
+
+protocol StakingUnbondSetupViewModelState: StakingUnbondSetupUserInputHandler {
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift:25]: mixed up

File: fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    func setup()
+    func estimateFee(builderClosure: ExtrinsicBuilderClosure?)
+}
+
+protocol StakingUnbondSetupUserInputHandler {
+    func selectAmountPercentage(_ percentage: Float)
+    func updateAmount(_ amount: Decimal)
+}
+
+extension StakingUnbondSetupUserInputHandler {}
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/Flow/StakingUnbondSetupFlow.swift:73]: empty

File: fearless/Modules/Staking/StakingUnbondSetup/StakingUnbondSetupViewFactory.swift [hunk 10]
```diff
@@ -2,40 +2,45 @@ import Foundation
...
     ) -> StakingUnbondSetupViewProtocol? {
+        let wireframe = StakingUnbondSetupWireframe()
+
+        let dataValidatingFactory = StakingDataValidatingFactory(presentable: wireframe)
+
+        guard let container = createContainer(chainAsset: chainAsset, wallet: wallet, dataValidatingFactory: dataValidatingFactory, flow: flow) else {
+            return nil
+        }
+
         guard let interactor = createInteractor(
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/StakingUnbondSetupViewFactory.swift:21]: in one guard state plz

File: fearless/Modules/Staking/StakingUnbondSetup/View/StakingUnbondSetupLayout.swift [hunk 11]
```diff
@@ -59,6 +62,19 @@ final class StakingUnbondSetupLayout: UIView {
             make.bottom.leading.trailing.equalToSuperview()
         }
 
+        contentView.stackView.addArrangedSubview(collatorView)
+        contentView.stackView.addArrangedSubview(accountView)
+
+        collatorView.snp.makeConstraints { make in
+            make.leading.trailing.equalToSuperview().inset(UIConstants.horizontalInset)
+            make.height.equalTo(52)
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/View/StakingUnbondSetupLayout.swift:70]: constants

File: fearless/Modules/Staking/StakingUnbondSetup/View/StakingUnbondSetupLayout.swift [hunk 12]
```diff
@@ -59,6 +62,19 @@ final class StakingUnbondSetupLayout: UIView {
...
+        }
+
+        accountView.snp.makeConstraints { make in
+            make.leading.trailing.equalToSuperview().inset(UIConstants.horizontalInset)
+            make.height.equalTo(52)
```
Comment (DRadmir) [inline fearless/Modules/Staking/StakingUnbondSetup/View/StakingUnbondSetupLayout.swift:75]: constants

File: fearless/ru.lproj/Localizable.strings [hunk 13]
```diff
@@ -692,7 +704,39 @@ and depending on the network state, may not be selected to participate";
...
 "chain.selection.hide.zero.balances" = "Скрыть с пустым балансом";
+"staking.stake.with.selected.title" = "Stake with selected";
+"staking.filter.title.rewards.apr" = "Rewards (APR)";
+"staking.custom.collators.title" = "Select a collator";
+"parachain.staking.recommended.section.title" = "Collators suggested by algorithm";
+"staking.recommended.filter.minimum.bond" = "Having relevant minimum bond";
+"staking.recommended.collators.empty.text" = "Your stake amount is less than the minimum one of all collators suggested by the algorithm. Increase your stake amount or choose a collator manually";
+"parachain.staking.reward.info.max" = "Максимальный APR";
+"parachain.staking.reward.info.avg" = "Средний APR";
+"parachain.staking.recommended.list.title" = "Предложено %d";
```
Comment (DRadmir) [inline fearless/ru.lproj/Localizable.strings:742]: file with English text

**Review by DRadmir — CHANGES_REQUESTED** [review]

**Review by DRadmir — CHANGES_REQUESTED** [review]
Please study the comments

File: fearless/Common/Model/ParachainStakingScheduledRequest.swift [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+enum ParachainStakingDelegationAction: Decodable, Equatable {
+    case revoke(amount: BigUInt)
+    case decrease(amount: BigUInt)
+
+    init(from decoder: Decoder) throws {
+        var container = try decoder.unkeyedContainer()
+
+        let type = try container.decode(String.self)
+        let amountString = try container.decode(String.self)
+        let amount = BigUInt(stringLiteral: amountString)
```
[the remark this answers is not in the document]
Comment (bnsports) [reply fearless/Common/Model/ParachainStakingScheduledRequest.swift:14]: @DRadmir  what should I see here ?

File: fearless/Modules/Staking/SelectValidatorsFlow/CustomValidatorList/Flow/Relaychain/CustomValidatorListRelaychainViewModelState.swift [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+            return .relaychainExisting(
+                validatorList: selectedValidatorList.items,
+                maxTargets: maxTargets,
+                state: bonding
+            )
+        }
+    }
+
+    func selectValidatorsConfirmFlow() -> SelectValidatorsConfirmFlow? {
+        nil
```
[the remark this answers is not in the document]
Comment (bnsports) [reply fearless/Modules/Staking/SelectValidatorsFlow/CustomValidatorList/Flow/Relaychain/CustomValidatorListRelaychainViewModelState.swift:73]: yes, there are no transition to confirm flow in relaychain case

File: fearless/Modules/Staking/Services/RewardCalculatorService/ParachainRewardCalculatorService.swift [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        do {
+            let localKey = try LocalStorageKeyFactory().createFromStoragePath(
+                .totalIssuance,
+                chainId: chainAsset.chain.chainId
+            )
+
+            let totalIssuanceDataProvider = providerFactory.createStorageProvider(for: localKey)
+
+            let updateClosure: ([DataProviderChange<ChainStorageItem>]) -> Void = { [weak self] changes in
+                let finalValue: ChainStorageItem? = changes.reduce(nil) { _, item in
```
[the remark this answers is not in the document]
Comment (bnsports) [reply fearless/Modules/Staking/Services/RewardCalculatorService/ParachainRewardCalculatorService.swift:211]: why?

File: fearless/Modules/Staking/Services/RewardCalculatorService/RewardCalculatorEngine.swift [hunk 17]
```diff
@@ -63,6 +68,143 @@ enum RewardCalculatorEngineError: Error {
...
+            totalStaked,
+            precision: assetPrecision
+        ) ?? 0.0
+        self.eraDurationInSeconds = eraDurationInSeconds
+        self.commission = commission
+        self.collators = collators
+    }
+
+    private lazy var annualInflation: Decimal = {
+        0.025
```
[the remark this answers is not in the document]
Comment (bnsports) [reply fearless/Modules/Staking/Services/RewardCalculatorService/RewardCalculatorEngine.swift:111]: from documentation

File: fearless/ru.lproj/Localizable.strings (hunk 13 above)
Comment (bnsports) [reply fearless/ru.lproj/Localizable.strings:742]: it's okay, that means there is no translation in lokalise. we doesn't have 100% russian translation

**Review by Viscount88 — APPROVED** [review]

**Review by DRadmir — APPROVED** [review]

**Pull request merged**